lib/client/client.go

```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
 	if params.RouteToDatabase.ServiceName != "" {
 		key.DBTLSCerts[params.RouteToDatabase.ServiceName] = certs.TLS
 	}
+	return key, nil
+}
 
-	// save the cert to the local storage (~/.tsh usually):
-	_, err = localAgent.AddKey(key)
-	return trace.Wrap(err)
+// IssueUserCertsWithMFA generates a single-use certificate for the user.
+func (proxy *ProxyClient) IssueUserCertsWithMFA(ctx context.Context, params ReissueParams) (*Key, error) {
+	localAgent := proxy.teleportClient.LocalAgent()
+	key, err := localAgent.GetKey(WithKubeCerts(params.RouteToCluster))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cert, err := key.SSHCert()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tlsCert, err := key.TeleportTLSCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	rootClusterName, err := tlsca.ClusterName(tlsCert.Issuer)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if params.RouteToCluster == "" {
+		params.RouteToCluster = rootClusterName
+	}
+
+	// Note: always connect to root for getting new credentials.
+	clt, err := proxy.ConnectToCluster(ctx, params.RouteToCluster, true)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer clt.Close()
+	requiredCheck, err := clt.IsMFARequired(ctx, params.isMFARequiredRequest(proxy.hostLogin))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if !requiredCheck.Required {
+		log.Debug("MFA not required for access.")
+		// MFA is not required.
+		// SSH certs can be used without embedding the node name.
+		if params.usage() == proto.UserCertsRequest_SSH {
+			return key, nil
+		}
+		// All other targets need their name embedded in the cert for routing,
+		// fall back to non-MFA reissue.
+		return proxy.reissueUserCerts(ctx, params)
+	}
+
+	if len(params.AccessRequests) == 0 {
+		// Get the active access requests to include in the cert.
+		var activeRequests services.RequestIDs
+		rawRequests, ok := cert.Extensions[teleport.CertExtensionTeleportActiveRequests]
+		if ok {
+			if err := activeRequests.Unmarshal([]byte(rawRequests)); err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+		params.AccessRequests = activeRequests.AccessRequests
+	}
+
+	// Always connect to root for getting new credentials, but attempt to reuse
+	// the existing client if possible.
+	if params.RouteToCluster != rootClusterName {
```

Comment:
If the comment above about connecting to root is accurate, then this is probably not needed?